server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -692,6 +692,7 @@ public void onFailure(Exception e) {
                     e
                 );
                 replicaRequest.getRequest().onRetry();
+                replicaRequest.incRef();
```

Comment:
I think in all the new calls to `incRef()` in this PR we should be certain that it will succeed (i.e. it'd be a bug for it to fail because the request was already closed) so it'd be better to use `mustIncRef()` instead.